pangeo_forge_recipes/recipes/xarray_zarr.py

```diff
@@ -206,303 +682,178 @@ def _set_target_chunks(self):
         else:
             self._concat_dim_chunks = self._nitems_per_input * self.inputs_per_chunk
 
+    # Each stage of the recipe follows the same pattern:
+    # 1. A top-level function, e.g. `prepare_target`, that does the actual work.
+    # 2. A private property, e.g. `._prepare_target`, that builds a partially applied function,
+    #    accepting just the arguments needed (e.g. a chunk_key).
+    # 3. A public property, e.g. `.prepare_target`, that calls the partially applied function
+    #    with the provided arguments (e.g. a chunk_key)
+    # This ensures that the actual function objects shipped to and executed on
+    # workers do not contain any references to the `recipe` object itself, which is complicated
+    # to serialize.
+
+    @property
+    def _prepare_target(self):
+        return functools.partial(
+            prepare_target,
+            target=self.target,
+            target_chunks=self.target_chunks,
+            init_chunks=self._init_chunks,
+            concat_dim=self._concat_dim,
+            nitems_per_input=self._nitems_per_input,
+            file_pattern=self.file_pattern,
+            inputs_chunks=self._inputs_chunks,
+            cache_metadata=self._cache_metadata,
+            chunks_inputs=self._chunks_inputs,
+            xarray_concat_kwargs=self.xarray_concat_kwargs,
+            process_chunk=self.process_chunk,
+            input_cache=self.input_cache,
+            cache_inputs=self.cache_inputs,
+            copy_input_to_local_file=self.copy_input_to_local_file,
+            xarray_open_kwargs=self.xarray_open_kwargs,
+            delete_input_encoding=self.delete_input_encoding,
+            process_input=self.process_input,
+            metadata_cache=self.metadata_cache,
+        )
+
     @property  # type: ignore
     @closure
     def prepare_target(self) -> None:
-        if self.target is None:
-            raise ValueError("target is not set.")
-        try:
-            ds = self.open_target()
-            logger.info("Found an existing dataset in target")
-            logger.debug(f"{ds}")
+        return self._prepare_target()
+
+    @property
+    def _cache_input(self):
+        return functools.partial(
+            cache_input,
+            cache_inputs=self.cache_inputs,
+            input_cache=self.input_cache,
+            file_pattern=self.file_pattern,
```

Comment:
I am having a hard time understanding why these partial functions are "small" in a serialization sense, since each function in the graph essentially contains a `functools.partial`-wrapped version of all of the same attributes that are part of the recipe class? In particular, `file_pattern` and `target`s are all fairly complex / large objects themselves, which are now being curried into a single-argument function. How is that any better than having a method attached to a class?

Reply:
Just noting that my initial attempt to answer this question have failed. My hunch is that it's just easier to serialize these functions than it is to serialize the dataclass, but I don't I'd like to have a stronger justification than that. Pointing to the workflow in https://github.com/pangeo-forge/pangeo-forge-azure-bakery/issues/10 is some evidence, but it's pretty indirect.

I'll keep trying to come up with a clear answer.

Reply:
To be clear, having a forensic understanding of this is interesting academically but far less important than actually shipping code that works. 😁  So given limited time, I would focus on the checklist above (rather than digging deeper). 

Let me know if you want help on any aspects here. I have stopped working on #153 in the meantime.

Reply:
I'll take care of the first few items around `to_pipelines()` and will look into the outstanding metadata caching.

Reply:
I updated the docs and notebooks. I had some trouble running the notebook, but I think it was an issue with my local internet connection.